web-app/src/lib/db/services/job.service.ts

```diff
@@ -41,6 +228,154 @@ export async function getJobsByAgentId(
   return jobs;
 }
 
+export async function syncJobStatus(job: Job) {
+  const agent = await getAgentById(job.agentId);
+
+  if (!agent) {
+    throw new Error("Agent not found");
+  }
+  const paymentClient = getPaymentClient();
+
+  const purchase = await getPurchase({
+    client: paymentClient,
+    query: {
+      cursorId: job.paymentId,
+      network: "Preprod",
+      limit: 1,
+    },
+  });
+
+  if (
+    purchase.error ||
+    !purchase.data ||
+    purchase.data.data.Purchases.length != 1
+  ) {
+    await prisma.job.update({
+      where: {
+        id: job.id,
+      },
+      data: {
+        status: "FAILED",
+        errorNote: purchase.error ? String(purchase.error) : "Unknown error",
+        errorNoteKey: "Job.SyncStatusFailed",
+      },
+    });
+    throw new Error("Failed to get on-chain status");
+  }
+
+  const onChainState = purchase.data.data.Purchases[0].onChainState;
+
+  if (onChainState === "FundsLocked") {
+    console.log("Funds still in locked state");
+    return;
+  }
+
+  if (onChainState === "ResultSubmitted" || onChainState == "Withdrawn") {
+    const baseUrl = agent.apiBaseUrl;
+    let baseUrlString = baseUrl.toString();
+    if (baseUrlString.endsWith("/")) {
+      baseUrlString = baseUrlString.slice(0, -1);
+    }
+    const syncJobUrl = new URL(
+      `${baseUrlString}/status?job_id=${job.agentJobId}`,
+    );
```

Comment:
Similar to the other `URL` comment. You have to do the slicing if we use the URL constructor and `URLSearchParams`.

```TypeScript
const baseUrl = 'https://example.com/';
const path = '/api/data';
const params = new URLSearchParams({
  userId: '123',
  type: 'full'
});

const url = new URL(path, baseUrl);
url.search = params.toString();

console.log(url.toString()); // Outputs: https://example.com/api/data?userId=123&type=full
```

Reply:
https://github.com/masumi-network/sokosumi/pull/208/commits/c737666945b528fe98185b993b49143e90821578